### PR TITLE
added authorized_keys table

### PR DESCRIPTION
**Pull request by PickmanSec**
2nd attempt at authorized keys table. Added comment ignoring and algorithm checking. Future work to be done is adding authorized_keys2 file and support for commands after successful login.

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by PickmanSec** [comment]
I ran make format but it didn't say anything about modifying any files.

File: osquery/tables/system/emacs [hunk 1]
```diff
@@ -0,0 +1 @@
+../Cellar/cmake/3.3.2/share/emacs
```
Comment (marpaia) [inline osquery/tables/system/emacs]: what is this?

**Comment by PickmanSec** [comment]
I am not sure where that came from but I removed it.

File: osquery/tables/system/emacs (hunk 1 above)
Comment (PickmanSec) [inline osquery/tables/system/emacs]: not sure

**Comment by PickmanSec** [comment]
I made fixes from PR https://github.com/facebook/osquery/pull/1616 but my commits got super messy and rather then try and make a bunch of commits I just made a new request.

**Comment by theopolis** [comment]
image

**Comment by theopolis** [comment]
Try this:
```
git reset --hard 4921432ae103b4a7c6ab9377cd7da14528c7dcca
git commit --amend
# Then delete all of the non-commented lines and replace with something like "Added authorized_keys table" or a description of what's added
git push -f
```

**Comment by PickmanSec** [comment]
Done, and yes that is what happened.

**Comment by theopolis** [comment]
Cool, that worked! Now that `osquery/tables/system/emacs` is back, so to remove it:
```
git rm osquery/tables/system/emacs
git commit --amend # this time don't change anything
git push -f
```

**Comment by theopolis** [comment]
Also, I've tested this on a few machines I own and I don't think it's following the `authorized_keys` specs completely, note the parts related to the format of each line [from the man page](http://man.he.net/man5/authorized_keys): 

```
     An example authorized_keys file:

        # Comments allowed at start of line
        ssh-rsa AAAAB3Nza...LiPk== user@example.net
        from="*.sales.example.net,!pc.sales.example.net" ssh-rsa
        AAAAB2...19Q== john@example.net

     added to the per-user file.

     Each line in these files contains the following fields: markers
     (optional), hostnames, bits, exponent, modulus, comment.  The fields are
     separated by spaces.

     The marker is optional, but if it is present then it must be one of
     ``@cert-authority'', to indicate that the line contains a certification
     authority (CA) key, or ``@revoked'', to indicate that the key contained
     on the line is revoked and must not ever be accepted.  Only one marker
     should be used on a key line.

     Hostnames is a comma-separated list of patterns ('*' and '?' act as wild-
     cards); each pattern in turn is matched against the canonical host name
     (when authenticating a client) or against the user-supplied name (when
     authenticating a server).  A pattern may also be preceded by '!' to indi-
     cate negation: if the host name matches a negated pattern, it is not
     accepted (by that line) even if it matched another pattern on the line.
     A hostname or address may optionally be enclosed within '[' and ']'
     brackets then followed by ':' and a non-standard port number.

     Alternately, hostnames may be stored in a hashed form which hides host
     names and addresses should the file's contents be disclosed.  Hashed
     hostnames start with a '|' character.  Only one hashed hostname may
     appear on a single line and none of the above negation or wildcard opera-
     tors may be applied.

     Bits, exponent, and modulus are taken directly from the RSA host key;
     they can be obtained, for example, from /etc/ssh/ssh_host_key.pub.  The
     optional comment field continues to the end of the line, and is not used.

     Lines starting with '#' and empty lines are ignored as comments.

     When performing host authentication, authentication is accepted if any
     matching line has the proper key; either one that matches exactly or, if
     the server has presented a certificate for authentication, the key of the
     certification authority that signed the certificate.  For a key to be
     trusted as a certification authority, it must use the ``@cert-authority''
     marker described above.

     The known hosts file also provides a facility to mark keys as revoked,
     for example when it is known that the associated private key has been
     stolen.  Revoked keys are specified by including the ``@revoked'' marker
     at the beginning of the key line, and are never accepted for authentica-
     tion or as certification authorities, but instead will produce a warning
     from ssh(1) when they are encountered.

     It is permissible (but not recommended) to have several lines or differ-
     ent host keys for the same names.  This will inevitably happen when short
     forms of host names from different domains are put in the file.  It is
     possible that the files contain conflicting information; authentication
     is accepted if valid information can be found from either file.

     Note that the lines in these files are typically hundreds of characters
        # A hashed hostname
        |1|JfKTdBh7rNbXkVAQCRp4OQoPfmI=|USECr3SWf1JUPsms5AqfD5QfxkM= ssh-rsa
        AAAA1234.....=
        # A revoked key
        @revoked * ssh-rsa AAAAB5W...
        # A CA key, accepted for any host in *.mydomain.com or *.mydomain.org
        @cert-authority *.mydomain.org,*.mydomain.com ssh-rsa AAAAB5W...
```

The hashed hostnames are very interesting! They are starting to show up inside `known_hosts` too!

One, related, final thought: it would be cool to build a family of ssh tables. So maybe we can rename this to `ssh_authorized_keys` and then have a similar `ssh_known_hosts` or `ssh_config_options`, etc later on.

**Comment by PickmanSec** [comment]
I agree with the other tables being added. I am working on several other tables including a known host table. One that note I can add the other fields to auth_keys table.

**Comment by PickmanSec** [comment]
So the lines should be parsed but there are a incredibly high number of options in these files. To get the most out of the table for the moment. I believe that just not including commented lines is worth it. We can add parsing to the line later.

File: osquery/tables/system/authorized_keys.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace osquery {
+namespace tables {
+
+const std::vector<std::string> kSSHAuthorizedkeys = {
+    ".ssh/authorized_keys",".ssh/authorized_keys2"
+};
+
+void genSSHkeysForUser(const std::string& username,
+                            const std::string& directory,
```
Comment (theopolis) [inline osquery/tables/system/authorized_keys.cpp]: nitpick, To get well-formatted code use `git add .` then `make format` then `git commit --amend` when you make changes and are ready to push back to the PR with `git push -f`

File: osquery/tables/system/authorized_keys.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    std::string keys_content;
+    if (!readFile(keys_file, keys_content).ok()) {
+      // Cannot read a specific keys file.
+      continue;
+    }
+    //Protocol 1 public key consist of: options, bits, exponent, modulus, comment
+    //Protocol 2 public key consist of: options, keytype, base64-encoded key, comment.
+    for (const auto& line : split(keys_content, "\n")) {
+      if (line[0] != '#') {
```
Comment (theopolis) [inline osquery/tables/system/authorized_keys.cpp]: Problem here! Check to make sure the line is not empty: 
```cpp
if (!line.empty() && line[0] !+ '#') {
```

`osquery::split` will trim the line automatically, but you may get empty entries. ;)

**Comment by theopolis** [comment]
@PickmanSec, cool, I agree we can solve this iteratively. If you don't mind squashing your commits again into 1 and checking out the `line.empty()` suggestion, we can merge this.

**Comment by PickmanSec** [comment]
Done.

**Comment by PickmanSec** [comment]
This build won't work yet I need to fix the check empty lines.

**Comment by PickmanSec** [comment]
Should be good now.

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Pull request merged**